xmrstak/backend/amd/amd_gpu/opencl/cryptonight.cl

```diff
@@ -761,22 +749,37 @@ __kernel void JH(__global ulong *states, __global uint *BranchBuf, __global uint
 		{
 			ulong input[8];
 
-			if(i < 3)
-			{
-				for(int x = 0; x < 8; ++x) input[x] = (states[(i << 3) + x]);
-			}
-			else if(i == 3)
-			{
-				input[0] = (states[24]);
-				input[1] = 0x80UL;
-				for(int x = 2; x < 8; ++x) input[x] = 0x00UL;
-			}
-			else
-			{
-				input[7] = 0x4006000000000000UL;
-
-				for(int x = 0; x < 7; ++x) input[x] = 0x00UL;
-			}
+			/* next code is equal to this block
+			 * if(i < 3)
+			 * {
+			 *     for(int x = 0; x < 8; ++x) input[x] = (states[(i << 3) + x]);
+			 * }
+			 * else if(i == 3)
+			 * {
+			 *     input[0] = (states[24]);
+			 *     input[1] = 0x80UL;
+			 *     for(int x = 2; x < 8; ++x) input[x] = 0x00UL;
+			 * }
+			 * else
+			 * {
+			 *     input[7] = 0x4006000000000000UL;
+			 *     for(int x = 0; x < 7; ++x) input[x] = 0x00UL;
+			 * }
+			 */
+
+			const bool isLessThree = i < 3;
+			const int shifted = i << 3;
+			input[0] = isLessThree ? (states[shifted]) : 0x00UL;
```

Comment:
Unless benchmarked  or otherwise confirmed to be optimised by the compiler, ?: operator is just another way to write if / else.

Instead:

Unroll the `for(int i = 0; i < 5; ++i)` loop - it is a stupid aggregate of different tasks anyway. The whole thing can be rewritten as:

2x

```
#pragma unroll 8
for(int x = 0; x < 8; ++x) input[x] = (states[(i << 3) + x]);
<epilogue - possibly inlined function>
```

1x

```
input[0] = (states[24]);
input[1] = 0x80UL;
#pragma unroll 6
for(int x = 2; x < 8; ++x) input[x] = 0x00UL;
<epilogue - possibly inlined function>
```

3x

````
input[7] = 0x4006000000000000UL;
#pragma unroll 7
for(int x = 0; x < 7; ++x) input[x] = 0x00UL;
<epilogue - possibly inlined function>
````

Reply:
No if is not equal to `cond ? :`. The short version is a select option. The if can be transformed into the short versions but it strongly depends on the compiler. The short version is translated into hardware select operations which are available on the gpus.

I do not understand what you mean with the three different code blocks. How should they avoid the branch divergence?

Reply:
Please keep in mind that unrolling loops with to high numbers (e.g. 8) will take a high number of registers and therefore can slow downthe kernel.

Reply:
@psychocrypt the code that i posted above can literally replace the `for(int i = 0; i < 5; ++i)`

Reply:
Yes I can join the hand unrilled code to a loop. I will do it later.